File: sorting/c/merge_sort.c
#include <stdlib.h>

void merge_sort(int* array, const int n);
void m_sort(int* array, int* temp, int start, int stop);
void merge(int* array, int* temp, int start, int middle, int stop);

void merge_sort(int* array, const int n)
{
	int* temp = malloc(sizeof(int) * n);
	m_sort(array, temp, 0, n);
	free( temp );
}

void m_sort(int* array, int* temp, int start, int stop)
{
	if(stop - start > 1)
	{
		int middle = start + (stop - start) / 2;
		m_sort(array, temp, start, middle);
		m_sort(array, temp, middle, stop);
		merge(array, temp, start, middle, stop);
	}
}

void merge(int* array, int* temp, int start, int middle, int stop)
{
	int i = start, j = middle, k = 0;
	while(i < middle && j < stop)
		if(array[i] < array[j])
			temp[ k++ ] = array[ i++ ];
		else
			temp[ k++ ] = array[ j++ ];

	while(i < middle)
		temp[ k++ ] = array[ i++ ];

	while(j < stop)
		temp[ k++ ] = array[ j++ ];

	for(i = 0; i < k; i++)
		array[start + i] = temp[i];
}

File: sorting/go/merge_sort.go
package sorting

func merge_sort(array []int) {
	var temp []int = make([]int, len(array))
	m_sort(array, 0, len(array), temp)
}

func m_sort(array []int, start int, end int, temp []int) {
	if end-start > 1 {
		var middle int = (start + end) / 2
		m_sort(array, start, middle, temp)
		m_sort(array, middle, end, temp)
		merge(array, start, middle, end, temp)
	}
}

func merge(array []int, start, middle, end int, temp []int) {
	var i, j, k int = start, middle, 0
	for i < middle && j < end {
		if array[i] < array[j] {
			temp[k] = array[i]
			i++
		} else {
			temp[k] = array[j]
			j++
		}
		k++
	}
	for i < middle {
		temp[k] = array[i]
		i++
		j++
	}
	for j < end {
		temp[k] = array[j]
		j++
		k++
	}
	for i = 0; i < k; i++ {
		array[start+i] = temp[i]
	}
}
